File: backend/app/services/audit_service.py
from datetime import datetime
from typing import Optional, Dict, Any, List
from sqlalchemy.orm import Session
from app.models.audit_log import AuditLog


class AuditService:
    @staticmethod
    def log_action(
        db: Session,
        action: str,
        resource_type: str,
        user_id: Optional[str] = None,
        resource_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> AuditLog:
        """
        Record a security/administrative action inside the audit log.
        """
        audit_entry = AuditLog(
            user_id=user_id,
            action=action,
            resource_type=resource_type,
            resource_id=resource_id,
            log_metadata=metadata
        )
        db.add(audit_entry)
        db.commit()
        db.refresh(audit_entry)
        return audit_entry

    @staticmethod
    def list_audit_logs(
        db: Session,
        user_id: Optional[str] = None,
        action: Optional[str] = None,
        resource_type: Optional[str] = None,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None,
        limit: int = 20,
        offset: int = 0
    ) -> List[AuditLog]:
        """
        Query audit log records with filtering options, ordered by timestamp DESC.
        """
        query = db.query(AuditLog)

        if user_id:
            query = query.filter(AuditLog.user_id == user_id)
        if action:
            query = query.filter(AuditLog.action == action)
        if resource_type:
            query = query.filter(AuditLog.resource_type == resource_type)
        if start_time:
            query = query.filter(AuditLog.timestamp >= start_time)
        if end_time:
            query = query.filter(AuditLog.timestamp <= end_time)

        return query.order_by(AuditLog.timestamp.desc()).offset(offset).limit(limit).all()

File: frontend/services/auditService.ts
import { fetchApi } from "./apiClient";
import { AuditLogRecord } from "../types/audit";

export interface AuditLogQueryParams {
  user_id?: string;
  action?: string;
  resource_type?: string;
  limit?: number;
  offset?: number;
}

export const auditService = {
  async getAuditLogs(params: AuditLogQueryParams = {}): Promise<AuditLogRecord[]> {
    const query = new URLSearchParams();
    if (params.user_id) query.append("user_id", params.user_id);
    if (params.action) query.append("action", params.action);
    if (params.resource_type) query.append("resource_type", params.resource_type);
    if (params.limit) query.append("limit", params.limit.toString());
    if (params.offset) query.append("offset", params.offset.toString());

    const qs = query.toString() ? `?${query.toString()}` : "";
    const rawLogs = await fetchApi<any[]>(`/audit-logs${qs}`);

    return rawLogs.map((log) => ({
      id: log.id,
      user_id: log.user_id,
      action: log.action,
      resource_type: log.resource_type,
      resource_id: log.resource_id,
      timestamp: log.timestamp,
      metadata: log.metadata || log.log_metadata || {},
    }));
  },
};
